base/sparse/symmetric.jl

```diff
@@ -0,0 +1,72 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+function Symmetric(A::SparseMatrixCSC, uplo::Symbol=:U)
+    checksquare(A)
+    Symmetric{eltype(A), typeof(A)}(A, Base.LinAlg.char_uplo(uplo))  # preserve A
+end
+
+(*)(A::Symmetric{TA,SparseMatrixCSC{TA,S}}, x::StridedVecOrMat{Tx}) where {TA,S,Tx} = A_mul_B(A, x)
+
+function A_mul_B!(α::Number, A::Symmetric{TA,SparseMatrixCSC{TA,S}}, B::StridedVecOrMat, β::Number, C::StridedVecOrMat) where {TA,S}
+    A.data.n == size(B, 1) || throw(DimensionMismatch())
+    A.data.m == size(C, 1) || throw(DimensionMismatch())
+    A.uplo == 'U' ? A_mul_B_U_kernel!(α, A, B, β, C) : A_mul_B_L_kernel!(α, A, B, β, C)
+end
+
+function A_mul_B(A::Symmetric{TA,SparseMatrixCSC{TA,S}}, x::StridedVector{Tx}) where {TA,S,Tx}
+    T = promote_type(TA, Tx)
+    A_mul_B!(one(T), A, x, zero(T), similar(x, T, A.data.n))
+end
+
+function A_mul_B(A::Symmetric{TA,SparseMatrixCSC{TA,S}}, B::StridedMatrix{Tx}) where {TA,S,Tx}
+    T = promote_type(TA, Tx)
+    A_mul_B!(one(T), A, B, zero(T), similar(B, T, (A.data.n, size(B, 2))))
+end
+
+function A_mul_B_U_kernel!(α::Number, A::Symmetric{TA,SparseMatrixCSC{TA,S}}, B::StridedVecOrMat, β::Number, C::StridedVecOrMat) where {TA,S}
```

Comment:
> There's a performance impact for that, which is why I separated them.

That should be compiled away.

> In fact, there's a performance impact for allowing `C` to be general, which makes me think that a separate version where `C` is just a vector is desirable.

If the impact is noticeable it makes sense to have a separate kernel for matvec.